# Udacity Project 3

# Gather

In [1]:
#imports that will be used for this file:
import pandas as pd
import requests
import os
import csv
import tweepy
import json
import re
import tweepy
import json
import time
import numpy as np
import math

In [2]:
df_arch = pd.read_csv("twitter-archive-enhanced.csv", sep= ",")
id_list = df_arch["tweet_id"]
#Testing with small batches below
#id_list = df_arch["tweet_id"][:100]
#id_list = [892420643555336193, 0]
#id_list = [892420643555336193, 892177421306343426, 0]

Image Prediction File Download

In [3]:
#download the file.
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
r = requests.get(url)


In [4]:
#save the .tsv file into the notebook folder. 
with open('file_2', 'w') as f:
    writer = csv.writer(f)
    reader = csv.reader(r.text.splitlines())
    
    for row in reader:
            writer.writerow(row)
#https://stackoverflow.com/questions/45978295/saving-a-downloaded-csv-file-using-python

In [5]:
#put it into a panda dataframe.
df_img_algo = pd.read_csv("file_2", sep= "\t")
df_img_algo.head()


tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

Third file from the twitter API

In [6]:
#consumer_key = 'MYKEY'
#consumer_secret = 'MYKEY'
#access_token = 'MYKEY-MYKEY'
#access_secret = 'MYKEY'
##file and list setup
#templist = []
#f = open("tweet_json.txt", "a+")
#count = 0
#auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
#auth.set_access_token(access_token, access_secret)
#api = tweepy.API(auth, wait_on_rate_limit= True, wait_on_rate_limit_notify= True)
#for id in id_list:
#    count += 1
#    try:
#        start = time.time()
#        tweet = api.get_status(id)
#        templist.append(tweet)
#        end = time.time()
#        print(count, "ID: ", id, "TIME: ", end - start)
#    except: 
#        print(count, "ID: ", id, " could not be found.")
#        continue

#tweet = api.get_status(666020888022790149)


In [7]:
#templist_json = []
#for item in templist:
#    templist_json.append(item._json)
#with open("tweet_json.txt", "w+") as f:
#    f.write(json.dumps(templist_json, indent = 4))
    
    

In [8]:
#create dataframe in Pandas. 
db_list = []

with open("tweet_json.txt") as f:
    json_temp = json.load(f)
    for item in json_temp:
        tweet_id = item["id"]
        tweet_text = item["text"]
        url = tweet_text[tweet_text.find("https"):]
        favorite_count = item["favorite_count"]
        retweet_count = item["retweet_count"] 
        hashtags = item["entities"]["hashtags"]
        followers_count = item["user"]["followers_count"]
        created_at = item["created_at"]
        
        db_list.append({"tweet_id": str(tweet_id),
                            "favorite_count": int(favorite_count),
                            "retweet_count": int(retweet_count),
                            "url": url,
                            "hashtags": hashtags,
                            "created_at": created_at,
                            "followers_count": followers_count
                            })
        
        df_tweet_json = pd.DataFrame(db_list, columns = ["tweet_id", "retweet_count", "favorite_count", "followers_count", 
                                                        "hashtags", "url", "created_at"])


In [9]:
#At this point we have three dataframes: df_arch, df_img_algo, df_tweet_json.
#We need to Assess the three data frames and clean them. 


# Assess

### Twitter .csv file

In [10]:
#First data type is the twitter archive which was lifted from the .csv file. 
df_arch.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

The df_arch dataframe has some noticible tidiness and quality issues. 
Firstly, the timestamp has a date and time. Ideally the date should be seperated out. Some quality issues I see visually are the conconsistencies between "null" data. None vs NaN can be unified into a "null" data. 

In [11]:
df_arch.info()
#This simple look up details tweet_id is incorrectly a integer. 
#It should be, like the zip code example we studied in the text, a string.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
df_arch.describe()
#This also reveals something erroneous about the data. THe denominator should be out of 10.
#But the mean is 10.45. It's also odd that the max is 2356.000000 and the min is 0.

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

# Issues with Twitter Archieve Data
### Tidiness
1)Date and Time should be seperated in "timestamp" or the datatype should be converted. 

2)Columns related to retweet should be dropped because there's not enough meaningful data.  

3)There should be a rating_value column that explains what the actual rating is in ratio. 

4) The dog names can be cobined into one. 
### Quality
1)href should be removed from the "source" column. It should just display the device. 

2)tweet_ID should be string. (like the zipcode example from the lsson)

3)Dog names have None. That should be converted into null. 

4)The reviewer pointed out that some ID's such had an incorrect rating lifted from the 'text' field. 

5)Rows that are retweets should be removed. 

# Image Data

In [13]:
#We can load the dataframe and take a look at the 
df_img_algo.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

## Quality Issue with Image Data: 
1)There isn't enough Hashtag Data. This should be removed entirely.
2)Tweet_id should be strings. 
3)Inconsistent capitalizations on "p" data. 
4)The "p" data is not descriptive enough.


# Tweets API Data


In [14]:
#visual examination first
df_tweet_json.head()

tweet_id  retweet_count  favorite_count  followers_count  \
0  892420643555336193           8642           38967          6551974   
1  892177421306343426           6351           33354          6551974   
2  891815181378084864           4213           25137          6551974   
3  891689557279858688           8759           42299          6551975   
4  891327558926688256           9521           40470          6551975   

  hashtags                      url                      created_at  
0       []  https://t.co/MgUWQ76dJU  Tue Aug 01 16:23:56 +0000 2017  
1       []  https://t.co/aQFSeaCu9L  Tue Aug 01 00:17:27 +0000 2017  
2       []  https://t.co/r0YlrsGCgy  Mon Jul 31 00:18:03 +0000 2017  
3       []  https://t.co/tD36da7qLQ  Sun Jul 30 15:58:51 +0000 2017  
4       []  https://t.co/0g0KMIVXZ3  Sat Jul 29 16:00:24 +0000 2017

In [15]:
#tweet_ID should be a string, and hashtags return a "[]" instead of Null. 
#also I am seeing that created_at time should be seperated out to date and time and day of the week.
df_tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2345 entries, 0 to 2344
Data columns (total 7 columns):
tweet_id           2345 non-null object
retweet_count      2345 non-null int64
favorite_count     2345 non-null int64
followers_count    2345 non-null int64
hashtags           2345 non-null object
url                2345 non-null object
created_at         2345 non-null object
dtypes: int64(3), object(4)
memory usage: 128.3+ KB


In [16]:
#the hashtag looks like there aren't many values. I created a simple loop to search
#for any legit dictionaries, There were only 19 enteries so I will delete this column. It's not very meaningful. 
count = 0
for item in df_tweet_json["hashtags"]:
    if len(item) > 0:
        count += 1
        print (item)
print ("total", count)

[{'text': 'BATP', 'indices': [21, 26]}]
[{'text': 'PrideMonthPuppo', 'indices': [29, 45]}]
[{'text': 'dogsatpollingstations', 'indices': [21, 43]}]
[{'text': 'PrideMonth', 'indices': [66, 77]}]
[{'text': 'PrideMonth', 'indices': [39, 50]}]
[{'text': 'ScienceMarch', 'indices': [36, 49]}]
[{'text': 'K9VeteransDay', 'indices': [40, 54]}]
[{'text': 'GoodDogs', 'indices': [62, 71]}]
[{'text': 'WKCDogShow', 'indices': [26, 37]}]
[{'text': 'BellLetsTalk', 'indices': [42, 55]}]
[{'text': 'WomensMarch', 'indices': [62, 74]}]
[{'text': 'ImWithThor', 'indices': [0, 11]}]
[{'text': 'BarkWeek', 'indices': [0, 9]}]
[{'text': 'BarkWeek', 'indices': [33, 42]}]
[{'text': 'BarkWeek', 'indices': [32, 41]}]
[{'text': 'FinalFur', 'indices': [40, 49]}]
[{'text': 'LoveTwitter', 'indices': [61, 73]}]
[{'text': 'notallpuppers', 'indices': [101, 115]}]
[{'text': 'NoDaysOff', 'indices': [98, 108]}, {'text': 'swole', 'indices': [109, 115]}]
total 19


# Issues with Twitter API Data: 

### Quality 
1)There isn't enough Hashtag Data. This should be removed entirely.

2)Tweet_id should be strings. 

### Tidiness Issues with Twitter API Data: 
1) created_at should either be parsed out to multiple fields or the data type should change to date. 

# Clean : Twitter Archive Data

In [17]:
#starting with copying df_arch
df_arch_clean = df_arch.copy()

### Issue
tweet_id should be strings, not integers. 


### Code

In [18]:
df_arch_clean["tweet_id"] = df_arch_clean["tweet_id"].astype("str")

### Test

In [19]:
type(df_arch_clean["tweet_id"])

pandas.core.series.Series

### Issue
Web Source should be just have the device, web ect. 


In [20]:
df_arch_clean["source"].value_counts()
#this explains that there are only four sources:iPhone, Vine, Twitter Web, and TweetDeck. 


<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

### Code

In [21]:
df_arch_clean['source'] = df_arch_clean['source'].str.replace('<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>', 'iPhone')
df_arch_clean['source'] = df_arch_clean['source'].str.replace('<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>', 'Vine')
df_arch_clean['source'] = df_arch_clean['source'].str.replace('<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>', 'Web')
df_arch_clean['source'] = df_arch_clean['source'].str.replace('<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>', 'TweetDeck')

In [22]:
df_arch_clean['source'] = df_arch_clean['source'].astype('category')

### Test

In [23]:
df_arch_clean['source'].value_counts()
#before the data source as category, it was int.

iPhone       2221
Vine           91
Web            33
TweetDeck      11
Name: source, dtype: int64

### Issue 

In [24]:
# retweets should be deleted. retweets can be discovered through running the below:
df_arch_clean[df_arch_clean['retweeted_status_timestamp'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
91    874434818259525634                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
109   871166179821445120                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
130   867072653475098625                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
171   858860390427611136                    NaN                  NaN   
180   857062103051644929                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
185   856330835276025856                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
...                  ...                    ...                  ...   
784   775096608509886464                    NaN                  NaN   
794   773336787167145985                    NaN                  NaN   
800   772615324260794368                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
815   771004394259247104                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
829   768909767477751808                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
841   766864461642756096                    NaN                  NaN   
847   766078092750233600                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
868   761750502866649088                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
911   757597904299253760                    NaN                  NaN   
926   754874841593970688                    NaN                  NaN   
937   753298634498793472                    NaN                  NaN   
943   752701944171524096                    NaN                  NaN   
949   752309394570878976                    NaN         

### Code

In [25]:
del_list = []
del_list = df_arch_clean[df_arch_clean['retweeted_status_timestamp'].notnull()]['tweet_id']
#this list contains the user id of the tweets that should be deleted.
del_index = del_list.index
#I can just drop the del index!
df_arch_clean = df_arch_clean.drop(del_index)


### Test

In [26]:
df_arch_clean[df_arch_clean['retweeted_status_timestamp'].notnull()]
#Got nothing back! it worked!

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

### Issue 
Retweet columns, and hashtag columns should be dropped. There's not enough on these columns:
in_reply_to_user_id
in_reply_to_status_id
retweeted_status_id
retweeted_status_user_id
retweeted_status_timestamp

In [27]:
#df_arch_clean['in_reply_to_user_id'].value_counts()
#df_arch_clean['in_reply_to_status_id'].value_counts()
#df_arch_clean['retweeted_status_id'].value_counts()
#df_arch_clean['retweeted_status_user_id'].value_counts()
#df_arch_clean['retweeted_status_timestamp'].value_counts()

### Code

In [28]:
df_arch_clean = df_arch_clean.drop('in_reply_to_user_id', axis = 1)
df_arch_clean = df_arch_clean.drop('in_reply_to_status_id', axis = 1)
df_arch_clean = df_arch_clean.drop('retweeted_status_id', axis = 1)
df_arch_clean = df_arch_clean.drop('retweeted_status_user_id', axis = 1)
df_arch_clean = df_arch_clean.drop('retweeted_status_timestamp', axis = 1)

### Test

In [29]:
df_arch_clean.info()
#columns are gone!

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null object
source                2175 non-null category
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: category(1), int64(2), object(9)
memory usage: 206.2+ KB


### Issue
Time stamp shows up with date and time. initlally I tried to seperate the time vs date by writing the seperate code below. I suppose you can create a seperate column for Date and Time, but there is a better way! I can just change the data format to "datetime64".


In [30]:
#parsing out date vs time in "timestamp":
date = []
time = []
for line in df_arch_clean["timestamp"]:
    n = line.split()[0]
    t = line.split()[1]
    date.append(n)
    time.append(t)
#printing out date and time will show that the two lists has the time and date separated. 

### Code

In [31]:
df_arch_clean.timestamp = pd.to_datetime(df_arch_clean.timestamp)

### Test

In [32]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null datetime64[ns]
source                2175 non-null category
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: category(1), datetime64[ns](1), int64(2), object(8)
memory usage: 206.2+ KB


### Issue

In [33]:
df_arch_clean.loc[df_arch_clean.tweet_id== '681340665377193984', 'text']
#some ratings were incorrectly extracted. For example ID: 681340665377193984 should be 9.5 but only 5 was picked up.

1689    I've been told there's a slight possibility he...
Name: text, dtype: object

### Code

In [34]:
#change the denominator and numerator as data type: floating point because these values are decimals
df_arch_clean['rating_denominator'] = df_arch_clean['rating_denominator'].astype('float')
df_arch_clean['rating_numerator'] = df_arch_clean['rating_numerator'].astype('float')


In [35]:
#view data from that has a decimal value from the "text". The reviewer pointed out that the the regular expression made an error and only extrate one number. 

df_arch_clean[df_arch_clean.text.str.contains(r"([0-9]+\.[0-9]*\/\d+)")]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  This is separate from the ipykernel package so we can avoid doing imports until


tweet_id           timestamp  source  \
45    883482846933004288 2017-07-08 00:28:19  iPhone   
695   786709082849828864 2016-10-13 23:23:56  iPhone   
763   778027034220126208 2016-09-20 00:24:34  iPhone   
1689  681340665377193984 2015-12-28 05:07:27  iPhone   
1712  680494726643068929 2015-12-25 21:06:00  iPhone   

                                                   text  \
45    This is Bella. She hopes her smile made you sm...   
695   This is Logan, the Chow who lived. He solemnly...   
763   This is Sophie. She's a Jubilant Bush Pupper. ...   
1689  I've been told there's a slight possibility he...   
1712  Here we have uncovered an entire battalion of ...   

                                          expanded_urls  rating_numerator  \
45    https://twitter.com/dog_rates/status/883482846...               5.0   
695   https://twitter.com/dog_rates/status/786709082...              75.0   
763   https://twitter.com/dog_rates/status/778027034...              27.0   
1689                                                NaN               5.0   
1712  https://twitter.com/dog_rates/status/680494726...              26.0   

      rating_denominator    name doggo floofer  pupper puppo  
45                  10.0   Bella  None    None    None  None  
695                 10.0   Logan  None    None    None  None  
763                 10.0  Sophie  None    None  pupper  None  
1689                10.0    None  None    None    None  None  
1712                10.0    None  None    None    None  None

These fields ID's are problematic.

883482846933004288  :Rating 13.5/10 

832215909146226688 :Rating 9.75/10

786709082849828864 :Rating 9.75/10

778027034220126208 :Rating 11.27/10

681340665377193984 :Rating 9.5/10

680494726643068929 :Rating 11.26/10

In [36]:
#df_arch_clean[df_arch_clean.tweet_id == '883482846933004288'].rating_numerator = 13.5
#this didn't work. 
#df_arch_clean.loc[df_arch_clean.tweet_id == '883482846933004288']=11.27
#this didn't work either. 
#http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy

df_arch_clean.loc[df_arch_clean.tweet_id == '883482846933004288', 'rating_numerator'] = 13.5
df_arch_clean.loc[df_arch_clean.tweet_id == '832215909146226688', 'rating_numerator'] = 9.75
df_arch_clean.loc[df_arch_clean.tweet_id == '786709082849828864', 'rating_numerator'] = 9.75
df_arch_clean.loc[df_arch_clean.tweet_id == '778027034220126208', 'rating_numerator'] = 11.27
df_arch_clean.loc[df_arch_clean.tweet_id == '681340665377193984', 'rating_numerator'] = 9.5
df_arch_clean.loc[df_arch_clean.tweet_id == '680494726643068929', 'rating_numerator'] = 11.26

### Test

In [37]:
df_arch_clean[df_arch_clean['tweet_id']== '883482846933004288'].rating_numerator
#worked

45    13.5
Name: rating_numerator, dtype: float64

### Issue
Rating as numerator/denominator is missing. 

### Code

In [38]:
df_arch_clean['rating_value'] = df_arch_clean['rating_numerator'] / df_arch_clean['rating_denominator']

### Test

In [39]:
df_arch_clean['rating_value'].head()
#It worked!

0    1.3
1    1.3
2    1.2
3    1.3
4    1.2
Name: rating_value, dtype: float64

### Issue 
The following columns have None instead of Null:
name doggo floofer pupper puppo

This column can be combined as one dog_stage

In [40]:
#df_arch_clean[df_arch_clean['doggo'] == 'None']
#df_arch_clean.doggo.head()

In [41]:
#replace 'None' with np.NaN
#https://stackoverflow.com/questions/34794067/how-to-set-a-cell-to-nan-in-a-pandas-dataframe
df_arch_clean.loc[df_arch_clean.doggo=='None', 'doggo'] = np.NaN
df_arch_clean.loc[df_arch_clean.floofer=='None', 'floofer'] = np.NaN
df_arch_clean.loc[df_arch_clean.pupper=='None', 'pupper'] = np.NaN
df_arch_clean.loc[df_arch_clean.puppo=='None', 'puppo'] = np.NaN


In [42]:
df_arch_clean[df_arch_clean['doggo'] == 'None']
#nothing shows up!

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, rating_value]
Index: []

In [43]:
df_arch_clean['dog_stage'] = df_arch_clean['text'].str.extract('(doggo|floofer|pupper|puppo)', expand=True)
#tried using melt and didn't work. Value got doubled...so I had to get creative with strings. 

In [44]:
#get rid of the columns
df_arch_clean = df_arch_clean.drop('doggo', axis=1)
df_arch_clean = df_arch_clean.drop('floofer', axis=1)
df_arch_clean = df_arch_clean.drop('pupper', axis=1)
df_arch_clean = df_arch_clean.drop('puppo', axis=1)

In [45]:
# and we need to use the 'melt' to combine these into a single column
#df_arch_clean['dog_stage'] = np.NaN
#df_arch_clean = pd.melt(df_arch_clean, 
#                        id_vars=['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
#       'rating_numerator', 'rating_denominator', 'name', 'rating_value'],
#                        var_name= 'dog_stage',
#                        value_name= 'temp')
#df_arch_clean = df_arch_clean.drop('temp', axis=1)


In [46]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 10 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null datetime64[ns]
source                2175 non-null category
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null float64
rating_denominator    2175 non-null float64
name                  2175 non-null object
rating_value          2175 non-null float64
dog_stage             364 non-null object
dtypes: category(1), datetime64[ns](1), float64(3), object(5)
memory usage: 172.2+ KB


# Clean: Image Data

In [47]:
#next we move on to the df_image_algo_clean data
#first, copy the data. 
df_img_algo_clean = df_img_algo.copy()
df_img_algo_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

### Issue
tweet_id should be strings

### Code 

In [48]:
df_img_algo_clean.tweet_id = df_img_algo_clean.tweet_id.astype(str)

### Test

In [49]:
df_img_algo_clean.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


### Issue
The following columns are not descriptive:
p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, 

### Code

In [50]:
df_img_algo_clean.rename(columns={'p1' : 'prediction_1',
                                  'p2' : 'prediction_2',
                                  'p3' : 'prediction_3',
                                  'p1_conf' : 'p1_confidence',
                                  'p2_conf' : 'p2_confidence',
                                  'p3_conf' : 'p3_confidence',
                                  'p1_dog' : 'predict_1_breed',
                                  'p2_dog' : 'predict_2_breed',
                                  'p3_dog' : 'predict_3_breed',
                                 }, inplace = True)
#https://chrisalbon.com/python/data_wrangling/pandas_rename_multiple_columns/

### Test

In [51]:
df_img_algo_clean.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'prediction_1', 'p1_confidence',
       'predict_1_breed', 'prediction_2', 'p2_confidence', 'predict_2_breed',
       'prediction_3', 'p3_confidence', 'predict_3_breed'],
      dtype='object')

### change values into lower case in the following columns: 
prediction_1
prediction_2
prediction_3

In [52]:
df_img_algo_clean['prediction_1'] = df_img_algo_clean['prediction_1'].str.lower()
df_img_algo_clean['prediction_2'] = df_img_algo_clean['prediction_2'].str.lower()
df_img_algo_clean['prediction_3'] = df_img_algo_clean['prediction_3'].str.lower()

In [53]:
df_img_algo_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num            prediction_1  p1_confidence  predict_1_breed  \
0        1  welsh_springer_spaniel       0.465074             True   
1        1                 redbone       0.506826             True   
2        1         german_shepherd       0.596461             True   
3        1     rhodesian_ridgeback       0.408143             True   
4        1      miniature_pinscher       0.560311             True   

         prediction_2  p2_confidence  predict_2_breed         prediction_3  \
0              collie       0.156665             True    shetland_sheepdog   
1  miniature_pinscher       0.074192             True  rhodesian_ridgeback   
2            malinois       0.138584             True           bloodhound   
3             redbone       0.360687             True   miniature_pinscher   
4          rottweiler       0.243682             True             doberman   

   p3_confidence  predict_3_breed  
0       0.061428             True  
1       0.072010             True  
2       0.116197             True  
3       0.222752             True  
4       0.154629             True

# Clean: Twitter JSON Data

In [54]:
#lastly we move on to df_tweet_json. 
df_tweet_json_clean = df_tweet_json.copy()

### Issue
There isn't enough data on hashtag column. We need to remove them.

### Code

In [55]:
df_tweet_json_clean = df_tweet_json_clean.drop('hashtags', axis = 1)

### Test

In [56]:
df_tweet_json_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2345 entries, 0 to 2344
Data columns (total 6 columns):
tweet_id           2345 non-null object
retweet_count      2345 non-null int64
favorite_count     2345 non-null int64
followers_count    2345 non-null int64
url                2345 non-null object
created_at         2345 non-null object
dtypes: int64(3), object(3)
memory usage: 110.0+ KB


### Issue
the created_at column has the wrong type. 

### Code

In [57]:
#df_tweet_json_clean['created_at'].astype(datetime)
df_tweet_json_clean.created_at = pd.to_datetime(df_tweet_json_clean.created_at)

### Test

In [58]:
df_tweet_json_clean['created_at'].head()
#looks great

0   2017-08-01 16:23:56
1   2017-08-01 00:17:27
2   2017-07-31 00:18:03
3   2017-07-30 15:58:51
4   2017-07-29 16:00:24
Name: created_at, dtype: datetime64[ns]

### Saving the files to CSV

In [59]:
df_arch_clean.to_csv('twitter_archive_master.csv')
df_img_algo_clean.to_csv('df_img_algo_clean')
df_tweet_json_clean.to_csv('df_tweet_json_clean')


In [60]:
#df_arch_clean.head(50)
#df_img_algo_clean.head(50)
#df_tweet_json_clean.head(50)

# Analysis 1)

When do people tweet? 

In [61]:
df1 = df_arch_clean.copy()
df1['timestamp'] = pd.to_datetime(df1['timestamp'])
df1['day_of_week'] = df1['timestamp'].dt.weekday_name
#https://stackoverflow.com/questions/30222533/create-a-day-of-week-column-in-a-pandas-dataframe-using-python

In [62]:
df1['day_of_week'].value_counts()



Monday       357
Tuesday      326
Wednesday    322
Thursday     305
Friday       305
Saturday     284
Sunday       276
Name: day_of_week, dtype: int64

In [63]:
df1[df1['day_of_week']== 'Monday']['tweet_id'].count()

357

In [64]:
import plotly
plotly.tools.set_credentials_file(username='
import plotly.plotly as py
#from plotly.graph_objs import *
import plotly.graph_objs as go

data = [go.Bar(
            x=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
            y=[df1[df1['day_of_week']== 'Monday']['tweet_id'].count(), 
               df1[df1['day_of_week']== 'Tuesday']['tweet_id'].count(), 
               df1[df1['day_of_week']== 'Wednesday']['tweet_id'].count(), 
               df1[df1['day_of_week']== 'Thursday']['tweet_id'].count(), 
               df1[df1['day_of_week']== 'Friday']['tweet_id'].count(), 
               df1[df1['day_of_week']== 'Saturday']['tweet_id'].count(), 
               df1[df1['day_of_week']== 'Sunday']['tweet_id'].count()
              ]
    )]
layout = go.Layout(  
    title='Day of the Week by number of Tweet'
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Day of the Week by number of Tweet')


I wanted to see which day of the week people tweet the most and the trend is insightful. It starts with Monday as the highest count and gradually declines over the days!

Monday       384
Tuesday      356
Wednesday    354
Friday       333
Thursday     326
Saturday     309
Sunday       294

# Analysis 2)

What were the inaccurate results for all three models? 
How accurate is the predictor model? 


In [65]:
df2= df_img_algo_clean.copy()
df2.info()
#df2
df2[(df2['predict_1_breed'] == False) & (df2['predict_2_breed'] == False) & (df2['predict_3_breed'] == False)]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id           2075 non-null object
jpg_url            2075 non-null object
img_num            2075 non-null int64
prediction_1       2075 non-null object
p1_confidence      2075 non-null float64
predict_1_breed    2075 non-null bool
prediction_2       2075 non-null object
p2_confidence      2075 non-null float64
predict_2_breed    2075 non-null bool
prediction_3       2075 non-null object
p3_confidence      2075 non-null float64
predict_3_breed    2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
50    666837028449972224    https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51    666983947667116034    https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
56    667065535570550784    https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
69    667188689915760640    https://pbs.twimg.com/media/CUJUk2iWUAAVtOv.jpg   
73    667369227918143488    https://pbs.twimg.com/media/CUL4xR9UkAEdlJ6.jpg   
77    667437278097252352    https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg   
78    667443425659232256    https://pbs.twimg.com/media/CUM8QZwW4AAVsBl.jpg   
93    667549055577362432    https://pbs.twimg.com/media/CUOcVCwWsAERUKY.jpg   
94    667550882905632768    https://pbs.twimg.com/media/CUObvUJVEAAnYPF.jpg   
96    667724302356258817    https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   
98    667766675769573376    https://pbs.twimg.com/media/CURiQMnUAAAPT2M.jpg   
100   667782464991965184    https://pbs.twimg.com/media/CURwm3cUkAARcO6.jpg   
106   667866724293877760    https://pbs.twimg.com/media/CUS9PlUWwAANeAD.jpg   
107   667873844930215936    https://pbs.twimg.com/media/CUTDtyGXIAARxus.jpg   
112   667911425562669056    https://pbs.twimg.com/media/CUTl5m1WUAAabZG.jpg   
115   667937095915278337    https://pbs.twimg.com/media/CUT9PuQWwAABQv7.jpg   
117   668142349051129856    https://pbs.twimg.com/media/CUW37BzWsAAlJlN.jpg   
118   668154635664932864    https://pbs.twimg.com/media/CUXDGR2WcAAUQKz.jpg   
123   668226093875376128    https://pbs.twimg.com/media/CUYEFlQXAAUkPGm.jpg   
130   668291999406125056    https://pbs.twimg.com/media/CUZABzGW4AE5F0k.jpg   
132   668466899341221888    https://pbs.twimg.com/media/CUbfGbbWoAApZth.jpg   
140   668544745690562560    https://pbs.twimg.com/media/CUcl5jeWsAA6ufS.jpg   
...                  ...                                                ...   
1839  837482249356513284    https://pbs.twimg.com/media/C59VqMUXEAAzldG.jpg   
1844  838916489579200512    https://pbs.twimg.com/media/C6RkiQZUsAAM4R4.jpg   
1847  839290600511926273    https://pbs.twimg.com/media/C6XBt9XXEAEEW9U.jpg   
1851  840370681858686976    https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg   
1853  840696689258311684    https://pbs.twimg.com/media/C6rBLenU0AAr8MN.jpg   
1869  844580511645339650    https://pbs.twimg.com/media/C7iNfq1W0AAcbsR.jpg   
1886  847962785489326080    https://pbs.twimg.com/media/C8SRpHNUIAARB3j.jpg   
1887  847971574464610304    https://pbs.twimg.com/media/C8SZH1EWAAAIRRF.jpg   
1891  849051919805034497    https://pbs.twimg.com/media/C8hwNxbXYAAwyVG.jpg   
1892  849336543269576704    https://pbs.twimg.com/media/C8lzFC4XcAAQxB4.jpg   
1900  851464819735769094    https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
1902  851861385021730816    https://pbs.twimg.com/media/C8W6sY_W0AEmttW.jpg   
1905  852226086759018497  https://pbs.twimg.com/ext_tw_video_thumb/85222...   
1906  852311364735569921    https://pbs.twimg.com/media/C9QEqZ7XYAIR7fS.jpg   
1910  853299958564483072    https://pbs.twimg.com/media/C9eHyF7XgAAOxPM.jpg   
1931  859074603037188101    https://pbs.twimg.com/media/C-wLyufW0AA546I.jpg   
1936  860184849394610176    https://pbs.twimg.com/media/C-_9jWWUwAAnwkd.jpg   
1937  860276583193509888    https://pbs.twimg.com/media/C_BQ_NlVwAAgYGD.jpg   
1940  860924035999428608    https://pbs.twimg.com/media/C_KVJjDXsA

There are 324 values that are false for all three predictions out of 2074. it's rougly about 15% completely incorrect. When the model is inaccruate, it predicts something other than a dog. 

In [66]:
df2['tweet_id'][(df2['predict_1_breed'] == True)].count()/2075
#model 1 is accurate .74 out of 1

0.7383132530120482

In [67]:
df2['tweet_id'][(df2['predict_2_breed'] == True)].count()/2075
#model 2 is accurate .75 out of 1

0.74843373493975907

In [68]:
df2['tweet_id'][(df2['predict_3_breed'] == True)].count()/2075
#model 3 is accurate .72 out of 1

0.72240963855421692

# Analysis 3)

How is WeRateDogs doing? have they been gaining followers? 

In [69]:
df3 = df_tweet_json_clean.copy()
#f3.columns
df3.sort_values(by = ['created_at'])

#Yes, they have been 


tweet_id  retweet_count  favorite_count  followers_count  \
2344  666020888022790149            521            2560          6552609   
2343  666029285002620928             47             131          6552609   
2342  666033412701032449             45             125          6552609   
2341  666044226329800704            141             302          6552609   
2340  666049248165822465             40             109          6552609   
2339  666050758794694657             58             133          6552609   
2338  666051853826850816            859            1229          6552609   
2337  666055525042405380            257             441          6552609   
2336  666057090499244032            142             300          6552609   
2335  666058600524156928             58             114          6552608   
2334  666063827256086533            223             480          6552608   
2333  666071193221509120             62             148          6552608   
2332  666073100786774016            169             326          6552608   
2331  666082916733198337             45             119          6552608   
2330  666094000022159362             76             164          6552608   
2329  666099513787052032             71             157          6552608   
2328  666102155909144576             13              80          6552608   
2327  666104133288665088           6696           14461          6552608   
2326  666268910803644416             36             105          6552608   
2325  666273097616637952             77             178          6552608   
2324  666287406224695296             68             151          6552608   
2323  666293911632134144            359             513          6552608   
2322  666337882303524864             94             200          6552608   
2321  666345417576210432            142             301          6552607   
2320  666353288456101888             74             224          6552607   
2319  666362758909284353            579             785          6552607   
2318  666373753744588802             95             191          6552607   
2317  666396247373291520             87             168          6552607   
2316  666407126856765440             41             110          6552607   
2315  666411507551481857            333             450          6552303   
...                  ...            ...             ...              ...   
29    886267009285017600              4             116          6551976   
28    886366144734445568           3226           21208          6551976   
27    886680336477933568           4531           22525          6551976   
26    886736880519319552           3341           12124          6551976   
25    886983233522544640           7868           35302          6551976   
24    887101392804085760           6040           30678          6551976   
23    887343217045368832          10537           33826          6551976   
22    887473957103951883          18468           69406          6551976   
21    887517139158093824          11810           46429          6551976   
20    887705289381826560           5464           30313          6551976   
19    888078434458587136           3547           21852          6551976   
18    888554962724278272           3643           20010          6551976   
17    888804989199671297           4411           25698          6551976   
16    888917238123831296           4564           29203          6551976   
15    889278841981685760           5516           25399          6551976   
14    889531135344209921           2262           15146          6551976   
13    889638837579907072           4600           27286          6551976   
12    889665388333682689          10193           48277          6551976   
11    889880896479866881           5030           27902          6551976   
10    890006608113172480           7439           30775          6551976   
9     890240255349198849           7524           32075         

In [70]:
#df3.columns
import plotly.plotly as py
import plotly.graph_objs as go

df3.drop('tweet_id', axis = 1)
df3.drop('url', axis = 1)
df3.sort_values(by= ['created_at'])



tweet_id  retweet_count  favorite_count  followers_count  \
2344  666020888022790149            521            2560          6552609   
2343  666029285002620928             47             131          6552609   
2342  666033412701032449             45             125          6552609   
2341  666044226329800704            141             302          6552609   
2340  666049248165822465             40             109          6552609   
2339  666050758794694657             58             133          6552609   
2338  666051853826850816            859            1229          6552609   
2337  666055525042405380            257             441          6552609   
2336  666057090499244032            142             300          6552609   
2335  666058600524156928             58             114          6552608   
2334  666063827256086533            223             480          6552608   
2333  666071193221509120             62             148          6552608   
2332  666073100786774016            169             326          6552608   
2331  666082916733198337             45             119          6552608   
2330  666094000022159362             76             164          6552608   
2329  666099513787052032             71             157          6552608   
2328  666102155909144576             13              80          6552608   
2327  666104133288665088           6696           14461          6552608   
2326  666268910803644416             36             105          6552608   
2325  666273097616637952             77             178          6552608   
2324  666287406224695296             68             151          6552608   
2323  666293911632134144            359             513          6552608   
2322  666337882303524864             94             200          6552608   
2321  666345417576210432            142             301          6552607   
2320  666353288456101888             74             224          6552607   
2319  666362758909284353            579             785          6552607   
2318  666373753744588802             95             191          6552607   
2317  666396247373291520             87             168          6552607   
2316  666407126856765440             41             110          6552607   
2315  666411507551481857            333             450          6552303   
...                  ...            ...             ...              ...   
29    886267009285017600              4             116          6551976   
28    886366144734445568           3226           21208          6551976   
27    886680336477933568           4531           22525          6551976   
26    886736880519319552           3341           12124          6551976   
25    886983233522544640           7868           35302          6551976   
24    887101392804085760           6040           30678          6551976   
23    887343217045368832          10537           33826          6551976   
22    887473957103951883          18468           69406          6551976   
21    887517139158093824          11810           46429          6551976   
20    887705289381826560           5464           30313          6551976   
19    888078434458587136           3547           21852          6551976   
18    888554962724278272           3643           20010          6551976   
17    888804989199671297           4411           25698          6551976   
16    888917238123831296           4564           29203          6551976   
15    889278841981685760           5516           25399          6551976   
14    889531135344209921           2262           15146          6551976   
13    889638837579907072           4600           27286          6551976   
12    889665388333682689          10193           48277          6551976   
11    889880896479866881           5030           27902          6551976   
10    890006608113172480           7439           30775          6551976   
9     890240255349198849           7524           32075         

In [71]:
data = [go.Scatter( x=df3['created_at'], y=df3['followers_count'] )]

layout = go.Layout(
    title='Day of the Week by number of Tweet'
)

fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Number of Followers by Year')
#this account has been losing followers


In [72]:
df_arch.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [73]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 10 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null datetime64[ns]
source                2175 non-null category
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null float64
rating_denominator    2175 non-null float64
name                  2175 non-null object
rating_value          2175 non-null float64
dog_stage             364 non-null object
dtypes: category(1), datetime64[ns](1), float64(3), object(5)
memory usage: 172.2+ KB
